src/edgepi/calibration/edgepi_eeprom.py

```diff
@@ -171,3 +181,128 @@ def page_write_register(self, mem_addr: int = None, data: list = None):
+        self.__parameter_sanity_check(start_addrx, length, True)
+        dummy_data = self.__generate_list_of_pages(start_addrx, [0]*length)
+        start_addrx = start_addrx + EdgePiMemoryInfo.USER_SPACE_START_BYTE.value
+        address_offset = 0
+        data_read = []
+        for data in dummy_data:
+            data_read = data_read + self.__sequential_read(start_addrx+address_offset, len(data))
+            address_offset = len(data)
+        return data_read
+
+    def write_memory(self, data: bytes):
+        """
+        Writes data to the eeprom
+        Args:
+            data (bytes): serialized json data
+        Return:
+            N/A
+        """
+        data_serialized = json.dumps(self.data_list) + data
+        mem_start = EdgePiMemoryInfo.USER_SPACE_START_BYTE.value + \
+            EdgePiMemoryInfo.BUFF_START.value
+        self.__parameter_sanity_check(mem_start, len(data_serialized), True)
+        pages_list = self.__generate_list_of_pages(mem_start, list(data_serialized))
+        self.__page_write_register(mem_start, pages_list)
+
+    def init_memory(self):
+        """
+        Initial Memory Reading
+        Args:
+            N/A
+        Return:
+            data_struct = list of data_struct
+        """
+        is_full=False
+        is_empty=False
+        mem_content = []
+        mem_size = self.__allocated_memory(EdgePiMemoryInfo.USER_SPACE_START_BYTE.value)
+
+        # mem_size of greater than EdgePiMemoryInfo.USER_SPACE_MAX.value should never happen
+        if mem_size > EdgePiMemoryInfo.USER_SPACE_MAX.value and mem_size != 0xFFFF:
+            raise ValueError(f'Invalid memory size read, possible data corruption, {mem_size}')
+
+        # Memory Full
+        if mem_size == EdgePiMemoryInfo.USER_SPACE_MAX.value and mem_size != 0xFFFF:
+            is_full=True
+            is_empty = False
+            self.log.warning('User Space Memory is full')
+            mem_start = EdgePiMemoryInfo.USER_SPACE_START_BYTE.value + \
+                        EdgePiMemoryInfo.BUFF_START.value
+            mem_content = bytes(self.read_memory(mem_start, mem_size))
+            self.data_list = json.loads(mem_content)
+            self.used_size = mem_size
+        # Memory Empty
+        elif mem_size == 0xFFFF:
+            is_full=False
+            is_empty=True
+            self.used_size = 0
+            self.log.info('User Space Memory is empty')
+        # part of memory occupied
+        else:
+            is_full=False
+            is_empty=False
+            mem_start = EdgePiMemoryInfo.USER_SPACE_START_BYTE.value + \
+                        EdgePiMemoryInfo.BUFF_START.value
+            # read memory content should be in json encoded bytes converted into list
+            mem_content = bytes(self.read_memory(mem_start, mem_size))
+            self.data_list = json.loads(mem_content)
+            self.used_size = mem_size
+            self.log.info(f'{mem_size}bytes of data is read from the user space')
```

Comment:
why is `is_empty` True if `mem_size` is 0xFFFF? is `mem_size` how much memory is taken? it looks like a big number to me so i thought it should be NOT empty

Reply:
I'm using the first two bytes of user space memory to represents the size of memory in use. When the memory is reset, all values in the memory are set to 0xFF which results in the mem_size being 0xFFFF